### ISSUE TITLE
Add generic recurrent block: .recur([...])

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#705]
## Problem

Recipe exposes fixed recurrent cells through `rnn(hidden)`, `gru(hidden)`, and `lstm(hidden)`, but it cannot make an arbitrary model composition recurrent.

A Rust `for` loop appends separately parameterized blocks. It does not reuse one block and its weights across sequence positions.

## Add

```rust
.recur([
	operations,
])
```

The array accepts any composable model content, including nested `res([...])` and `moe(top_k, [...])` blocks.

`recur([...])` declares the body once. At each sequence position, Recipe runs that same body with the current position input and the previous body's output as recurrent state. The body parameters are shared across every position. State starts at the existing zero-state boundary for each independent sequence.

Do not add an iteration-count argument. Training and prefill use the input sequence length. Incremental decode uses one recurrence per generated position and carries the state across calls. An outer Rust loop declares multiple stacked recurrent blocks, each with its own parameters:

```rust
for _ in 0..layers {
	model = model.recur([
		operations,
	]);
}
```

Keep fixed-depth repeated application separate from recurrence.

## Verify

- A recurrent body containing ordinary operations, `res([...])`, and `moe(top_k, [...])` lowers through the public Recipe API.
- The stored model contains one parameter set for the recurrent body, independent of sequence length.
- Forward and backward results match a hand-unrolled reference that reuses the same parameters.
- State carries between positions and resets between independent sequences.
- Stateful decode carries the same recurrent state across calls as one whole-sequence forward.
- A Rust loop around `recur([...])` creates the requested number of stacked recurrent blocks with distinct parameter sets.

Related: #77, #79, #566, and #573.

**Comment by nm-z** [comment]
No PR for this one yet, and I want to say why in terms of the code rather than leave it looking unattended. I opened PRs for the other seven issues in this batch (#704, #706–#711); this is the one I could not do honestly at this size.

## The tempting version is fake

`recur([layer(n)])` can be made to work today in about four lines: lower it to `lower_scan(graph, n, 1)`, which is exactly what `rnn(n)` already does. It would compile, train, save and reload, and it would pass any test I wrote for it.

It would also add no capability whatsoever — a new spelling for `rnn`. The issue asks for something else, and three separate things in the code block it.

## 1. `Primitive::Scan` is a cell, not a body

`lower_scan` (`recipe.rs:7095`) allocates `gates × (input matrix + state matrix + bias)` and pushes one node. `scan_forward_body` walks the sequence inside the kernel. There is no sub-graph parameter — the cell it applies per position is fixed arithmetic, and `gates` only picks how many of that same cell shape are stacked (1 for `rnn`, 3 for `gru`, 4 for `lstm`).

So "run this body at each position" has nowhere to go in `Scan`. It is not that `Scan` is close to general and needs a flag; the body is not data it can take.

## 2. Nodes cannot share parameters, and gradients would clobber

The other route is unrolling the body once per position with shared weights. Two things stop it.

`push_node` is the only node constructor and always allocates fresh (`recipe.rs:6771-6772`):

```rust
let (source, offset) = (graph.source, graph.parameters.len());
graph.parameters.resize(checked_add(offset, parameters, "model parameters")?, 0.0);
```

Nothing in the graph can point two nodes at one parameter range.

And if it could, the backward would be wrong. Contraction gradients go through split-K scratch and are folded by `reduce_rows`, which **seeds from the first scratch row and plain-stores the result** (`amd-nv-cpu.ll:106` and `:123`):

```llvm
%source.first = load double, ptr addrspace(1) %source.first.ptr, align 8   ; seed
...
store double %sum, ptr addrspace(1) %target.ptr, align 8                   ; overwrite
```

That is an overwrite, not an accumulate. Two nodes sharing an offset would have the later reduction discard the earlier one's gradient, so every position but one would contribute nothing. Shared parameters need an accumulate mode here plus an explicit zeroing pass, in every reverse path that writes gradients — not just this one.

## 3. There is no way to address one position

Even with sharing, unrolling needs position `t` of a `channels × length` tensor as a node input. The primitives are (`recipe.rs:6489`):

```rust
Contraction = 0, Pool = 2, Attention = 4, Scan = 5, Elementwise = 6, Normalize = 8, Predictor = 9,
```

None of them slices or gathers along `length`. So the unrolled form is not expressible as graph nodes either, with or without sharing.

## 4. The body type cannot hold what the issue asks it to hold

The issue says the array accepts nested `res([...])` and `moe(top_k, [...])`. `Residual` is three shallow variants (`recipe.rs:4802`):

```rust
pub enum Residual { Layer(usize), Conv(usize, usize), Activation(Activation) }
```

There is no variant for a nested block, and `Operation::Residual(Vec<Residual>)` / `Operation::Moe(usize, Vec<Residual>)` inherit that limit — `res` cannot contain `res` today either. A `recur` body that accepts them needs a deeper type and a bundle encoding for it, which is its own change with its own round-trip evidence. (`feat/569-hyper` carries a nested-block encoding via `text()`/`untext()` escaping; that is the closest existing prior art.)

## What this actually needs

In dependency order:

1. A nestable body type plus its record encoding, so `res`/`moe`/`recur` can contain each other. Independently useful — it fixes nested `res` too.
2. Gradient accumulation: an accumulate mode on `reduce_rows` and the other gradient writers, and an explicit zeroing pass, so more than one node can land on one parameter range.
3. Parameter sharing in the graph: a node offset that can point at an existing range, and the allocator/serializer agreeing on the parameter count.
4. Either a slice primitive along `length`, or a scan kernel that interprets a sub-graph per position. This is the new kernel and the largest piece.

Only then do the issue's Verify bullets mean anything — a hand-unrolled reference to match, state resetting between sequences, decode carrying state across calls.

I would rather leave this open and accurate than ship the four-line alias and call the issue closed. If a smaller first slice is wanted, item 1 is the one with standalone value and no kernel work, and it can be its own issue.

**Comment by nm-z** [comment]
Update after #732, and one correction to my earlier analysis.

## Blocker 1 is now gone

#732 makes a branch step an ordinary `Block`, so `res`/`moe` nest to any depth and a body type that can hold nested fragments exists. That was item 1 of the four I listed above, the one I said had standalone value and no kernel work.

## But there is a fifth blocker, and it rules out the unroll

I went back to design the unrolled form — slice position `t`, run the body, carry the state — because with #732 the nested-fragment requirement would come free: each step would lower through `lower_block`, exactly as the outer sequence does.

It does not work, for a reason not on my earlier list. **A node writes its own value arena and nothing else.** The layout is one arena per node index (`layout.values[index]`, `recipe.rs:1540`), and a node addresses at most two inputs, `source` and `second`. So `L` unrolled positions produce `L` separate length-1 state tensors and there is no way to assemble them into the `{channels, length}` output that `recur` has to return — no node can take `L` sources, and no node can write into a neighbour's arena. A concat would have to be a node with `L` inputs, which the `Node` shape cannot express.

That is independent of the gradient problem, so it is not worked around by fixing gradients.

## What that leaves

The in-kernel loop is the only form: one node holding the body's parameters once, whose kernel walks positions and evaluates the body per position, the way `scan_forward_body` walks time for a fixed cell (`amd-nv-cpu.ll:2357`). That kernel precomputes the input contraction over the whole sequence and then does only the state-dependent part per position, which is the shape a general `recur` wants too — but its cell is fixed, and a general body means interpreting a list of operations inside the loop.

So the revised list is:

1. ~~A nestable body type~~ — done in #732.
2. A `Primitive::Recur` whose kernel evaluates a body per position, forward and reverse. This is the whole of the work now. Parameter sharing and gradient accumulation, items 2 and 3 above, are **not** needed on this path: one node holds one parameter span and writes its gradient once, exactly as `Scan` does.

I no longer think items 2 and 3 are the way in. One node with a body-interpreting kernel is a smaller and safer change than sharing parameters across unrolled copies, because it touches no existing gradient writer — the contraction reverse path is register-tiled with lane-owner stores and split reductions (`amd-nv-cpu.ll:2518-2613`), and adding an accumulate mode there would change the gradient write for every model in the repository, with a silent failure mode.

## Not attempted

I have not written that kernel. It is the largest single piece in this issue and it needs a cross-backend byte-identity run plus a gradient check against a hand-unrolled reference to be worth anything, and archy — the machine those run on — is currently unreachable, no ICMP, since the driver fault that took dies 4 and 5 out of enumeration earlier today. I would rather leave this open with an accurate plan than land an unverified recurrent kernel.

**Comment by nm-z** [comment]
Measured, so the "it would add nothing" claim above is not just an assertion.

archy came back at 16:26 (it had rebooted; all eight dies enumerate again, including 4 and 5). I built the scaffolding — `Operation::Recur(Vec<Residual>)`, the `recur([...])` builder on `Model` and as a nested step, the record and its parse, and a `recur_cell` that reads a body as one parameterized step plus the activation that closes it — and lowered the one-stage case onto the existing scan node.

Then I compared it against `rnn`, three epochs, fp32, seed 17, CPU backend:

```
RNN    loss=19.24993232998686565 bits=40333ffb90af81b5
RECUR  loss=19.24993232998686565 bits=40333ffb90af81b5
loss identical        = true
predictions identical = true
```

`recipe.model().recur([layer(6), tanh()])` and `recipe.model().rnn(6)` are the same computation **to the bit** — same final loss, same every prediction. The saved bundles differ in exactly one line, the `block rnn,6` versus `block recur,…` record, and in nothing else.

So the one-stage form is `rnn` with a second spelling. I am not opening a PR for it: it would add public API surface that computes nothing new, and it would make this issue look addressed while `recur([layer(8), relu(), layer(8)])` and `recur([res([...])])` — the cases the issue is actually about — still cannot be written.

What that measurement does establish is that the body-to-cell reading works and the lowering path is sound. The remaining work is entirely the kernel: a `Primitive::Recur` whose forward walks positions evaluating a multi-stage body, and whose reverse carries the adjoint back through those stages and into the previous position. Everything above the kernel is a small change and I have it working.

The scaffolding is on a local branch, unpushed, since it is not worth landing on its own.

**Comment by nm-z** [comment]
Correcting myself again, and this one makes the work smaller. I said above that the remaining piece is "a kernel that interprets a body per position". That framing is wrong, and it made this look harder than it is.

**Recipe does not interpret at runtime — it emits IR per node, forward and reverse.** A node can already carry a program, and the emitter generates both directions from it:

- forward: `self.graph.programs.get(node.program_offset..code_end)` at `recipe.rs:2556`
- reverse: the same span read again at `recipe.rs:2697`, the "scalar reverse program"

So `ScalarProgram` is already a small language with a forward emitter and a reverse emitter, stored per node in `graph.programs` and turned into straight-line IR at emit time. That is exactly the shape `recur` needs, one level up.

## The revised plan

A `Primitive::Recur` node carries its body as a program in the same store, and the emitter writes the body inline inside the position loop:

- forward: emit, per row and per position, the body's operations as straight-line IR — a matvec for each contraction, the existing scalar-program emission for each activation, a per-row reduction for each normalization — with the state read from the previous position and written to this one. The `scan_forward_body` precompute trick still applies: the input-side contraction is done once over the whole sequence before the time loop, and only the state-dependent part runs per position.
- reverse: emit the adjoint of that same body, walking positions backwards, carrying the state adjoint into the previous position. The scalar reverse emitter is the prior art for the elementwise steps; the contraction steps need their matvec adjoint and a weight-gradient accumulation into the node's own single parameter span.

Nesting comes out of this for free, which is the part that matters for this issue's Verify list. `res([...])` inside the body is an add over emitted straight-line code, and `moe(top_k, [...])` is a gate over the same — both are more emitted IR, not new kernel entry points. That is only true because #732 made a branch step an ordinary `Block`, so the body is a real sub-graph rather than a restricted enum.

## Why this is better than what I proposed before

My earlier two plans were (1) unroll the body across positions in the graph, killed by blocker 5 — a node writes only its own arena, so `L` per-position states cannot be assembled; and (2) hand-write a large `Primitive::Recur` kernel with a fixed cell shape, which would not have supported nested fragments at all and so would not have closed this issue.

This third plan is an extension of a mechanism already in the tree, with a working forward and reverse emitter to model it on, and it is the first one that can actually satisfy the nested-body requirement.

## Still not done

I have not written it. The weight-gradient accumulation into one parameter span and the reverse emission for the contraction steps are the substantial parts, and they want a forward check against a reference recurrence computed from the saved weights, plus the cross-backend byte-identity run, before anyone should trust them. What is measured and standing is above: the scaffolding lowers, and the one-stage form is bit-identical to `rnn`.
